components/merkledb/src/migration/persistent_iter.rs

```diff
@@ -0,0 +1,669 @@
+/// // known position (the element with 0-based index 100, in our case).
+/// let mut iter = PersistentIter::new(helper.scratchpad(), "list_iter", &list);
+/// let (i, item) = iter.next().unwrap();
+/// assert_eq!(i, 100);
+/// assert_eq!(item, "100");
+/// assert_eq!(iter.count(), 22); // number of remaining items
+/// ```
+///
+/// [`Scratchpad`]: struct.Scratchpad.html
+/// [`MigrationHelper`]: struct.MigrationHelper.html
+pub struct PersistentIter<T: RawAccess, I: ContinueIterator> {
+    inner: Inner<T, I>,
+}
+
+impl<T, I> fmt::Debug for PersistentIter<T, I>
+where
+    T: RawAccess,
+    I: ContinueIterator,
+    I::Key: fmt::Debug,
+{
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter
+            .debug_struct("PersistentIter")
+            .field("inner", &self.inner)
+            .finish()
+    }
+}
+
+/// Internal details of a persistent iterator.
+enum Inner<T: RawAccess, I: ContinueIterator> {
+    /// The iterator is active: it has an underlying iterator over a database object,
+    /// and an entry storing the iterator position.
+    Active {
+        iter: Peekable<I::Iter>,
+        position_entry: Entry<T, IteratorPosition<I::Key>>,
+    },
+    /// The iterator has ended.
+    Ended,
+}
+
+impl<T, I> fmt::Debug for Inner<T, I>
+where
+    T: RawAccess,
+    I: ContinueIterator,
+    I::Key: fmt::Debug,
+{
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Inner::Active { position_entry, .. } => formatter
+                .debug_struct("Active")
+                .field("position", &position_entry.get())
+                .finish(),
+            Inner::Ended => formatter.debug_tuple("Ended").finish(),
+        }
+    }
+}
+
+impl<T, I> PersistentIter<T, I>
+where
+    T: RawAccessMut,
+    I: ContinueIterator,
+{
+    /// Creates a new persistent iterator.
+    pub fn new<A>(access: A, name: &str, index: I) -> Self
+    where
+        A: Access<Base = T>,
+    {
+        let position_entry: Entry<_, IteratorPosition<I::Key>> = access.get_entry(name);
+        let position = position_entry.get();
+
+        let start_key = match position {
+            None => None,
+            Some(IteratorPosition::NextKey(key)) => Some(key),
+            Some(IteratorPosition::Ended) => {
+                return Self {
+                    inner: Inner::Ended,
+                };
+            }
+        };
+
+        Self {
+            inner: Inner::Active {
+                iter: index
+                    .continue_iter(start_key.as_ref().map(Borrow::borrow))
+                    .peekable(),
+                position_entry,
+            },
+        }
+    }
+}
+
+impl<T, I> Iterator for PersistentIter<T, I>
+where
+    T: RawAccessMut,
+    I: ContinueIterator,
+{
+    type Item = <I::Iter as Iterator>::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.inner {
```

Comment:
Nit/style: Maybe `if let` syntax here will be a bit prettier -- usually it looks better in cases when one match arm is big, and other is very short. On the other hand, if one day a new variant will be added to `PersistentIter`, it will compile, while the current code will not (but it's unlikely, as far as I can see).